wallet/controllers_v3.go

```diff
@@ -0,0 +1,444 @@
+func LinkUpholdDepositAccountV3(s *Service) func(w http.ResponseWriter, r *http.Request) *handlers.AppError {
+	return func(w http.ResponseWriter, r *http.Request) *handlers.AppError {
+		var (
+			ctx = r.Context()
+			id  = new(inputs.ID)
+			cuw = new(LinkUpholdDepositAccountRequest)
+		)
+		// get logger from context
+		logger, err := appctx.GetLogger(ctx)
+		if err != nil {
+			// no logger, setup
+			ctx, logger = logging.SetupLogger(ctx)
+		}
+
+		// get payment id
+		if err := inputs.DecodeAndValidateString(context.Background(), id, chi.URLParam(r, "paymentID")); err != nil {
+			logger.Warn().Str("paymentID", err.Error()).Msg("failed to decode and validate paymentID from url")
+			return handlers.ValidationError(
+				"error validating paymentID url parameter",
+				map[string]interface{}{
+					"paymentID": err.Error(),
+				},
+			)
+		}
+
+		// read post body
+		if err := inputs.DecodeAndValidateReader(r.Context(), cuw, r.Body); err != nil {
+			return cuw.HandleErrors(err)
+		}
+
+		// get the wallet
+		wallet, err := s.GetWallet(id.UUID())
+		if err != nil {
+			if strings.Contains(err.Error(), "looking up wallet") {
+				return handlers.WrapError(err, "unable to find wallet", http.StatusNotFound)
+			}
+			return handlers.WrapError(err, "unable to get or create wallets", http.StatusServiceUnavailable)
+		}
+
+		var aa uuid.UUID
+
+		if cuw.AnonymousAddress != "" {
+			aa, err = uuid.FromString(cuw.AnonymousAddress)
+			if err != nil {
+				return handlers.ValidationError(
+					"error validating anonymousAddress",
+					map[string]interface{}{
+						"anonymousAddress": err.Error(),
+					},
+				)
+			}
+		}
+
+		publicKey, err := hex.DecodeString(wallet.PublicKey)
+		if err != nil {
+			logger.Warn().Err(err).Msg("unable to decode wallet public key")
+			return handlers.WrapError(errors.New("unable to decode wallet public key"),
+				"unable to decode wallet public key for creation request validation",
+				http.StatusInternalServerError)
+		}
+		uwallet := uphold.Wallet{
+			Info:    *wallet,
+			PrivKey: ed25519.PrivateKey{},
+			PubKey:  httpsignature.Ed25519PubKey([]byte(publicKey)),
+		}
+
+		// if this if we do not have a UserDepositAccountProviderID on this wallet,
+		// and it is a "brave" provided wallet
+		if wallet.UserDepositAccountProviderID == nil && wallet.Provider == "brave" {
+			// parse the signedlinkingreationrequest to get the provider id
+			txInfo, err := uwallet.VerifyTransaction(cuw.SignedLinkingRequest)
+			if err != nil {
+				logger.Warn().Err(err).Msg("failed to transaction validation for uphold")
+				return handlers.WrapError(
+					errors.New("unable to create uphold wallet"),
+					"failed transaction validation for uphold", http.StatusBadRequest)
+			}
+			logger.Debug().Msg("able to verify transaction")
+			// get the card id from the submitted destination
+			*wallet.ProviderLinkingID = uuid.Must(uuid.FromString(txInfo.Destination))
```

Comment:
this is different than how we've been deriving the linking id on bat-ledger I think

Reply:
fixed in: 351c7a7